Sources/MongoSwift/BSON/Codec.swift

```diff
@@ -333,3 +333,580 @@ private func getBsonValue(_ value: Any) -> BsonValue? {
         preconditionFailure("Value \(value) with type \(type(of: value)) didn't match any expected types")
     }
 }
+
+/// `BsonDecoder` facilitates the decoding of BSON into semantic `Decodable` types.
+public class BsonDecoder {
+
+    /// Contextual user-provided information for use during decoding.
+    open var userInfo: [CodingUserInfoKey: Any] = [:]
+
+    /// Options set on the top-level decoder to pass down the decoding hierarchy.
+    fileprivate struct _Options {
```

Comment:
unclear what we would use this or `userInfo` for right now, but eventually if we supported some different decoding strategies or something we could store that info in here. (and the `Decoder` protocol requires a `userInfo` property anyway)